enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,444 @@
+or some very general text/YAML substitution mechanism.
+
+There would still be a pod vs. secret creation race.
+
+Setting up a StatefulSet so that pods can’t impersonate each other
+would no longer require admins to manually use init containers;
+the init container approach is cumbersome, but ultimately probably secure enough.
+
+### Generate a Single Wildcard Certificate for the StatefulSet
+This would simplify the implementation, but the too broad subject introduces extra risk.
+
+https://github.com/kubernetes/dns/blob/master/docs/specification.md
+only specifies what DNS records must exist, not what must not exist,
+so such a certificate could match unwanted DNS entries.
+
+Notably once a wildcard certificate is created,
+it’s no longer possible to rely on TLS to prevent a compromised StatefulSet pod
+from impersonating a different pod from that StatefulSet.
+
+(The simple variant of just generating a `*.${service.name}…` without including the
+StatefulSet name at all also does not differentiate between different StatefulSets
+using the same headless service for hostnames.
+The more specific variant of `${statefulSet.name}-*.${service.name}…` only avoids this
+to an extent, as long as StatefulSet names are not prefixes of other StatefulSet names
+(`database` vs. `database-staging`),
+and there’s still the general downside of inability prevent impersonation within
+the StatefulSet.)
+
+### Annotate a Service Instead of a StatefulSet
+This would certainly be simpler
+for generating the very generic wildcard certificates (`*.${service.name}…`):
+they could be generated with the service, i.e. before creating the StatefulSet,
+and the pod vs. secret creation race would not exist.
+OTOH that only works well with such generic wildcards, which are undesirable.
+
+For generating individual pod certificates
+(or the more constrained `${statefulSet.name}-*.${service.name}…` wildcard variant),
+where the certificate subject is
+`${statefulSet.name}-${podID}.${service.name}.${service.namespace}.svc`,
+it is both more correct and simpler to annotate StatefulSets:
+
+- In the most general case,
+  there could be two or more different StatefulSets using the same headless service,
+  with one of the StatefulSets requesting a set of certificates,
+  and the other one not requesting certificates
+  (e.g. because they only communicate using shared storage, or some HW-specific mechanism).
+  So, conceptually it’s a better fit to place the certificate request on the StatefulSet
+  than on the service.
+- Implementation-wise, to generate the
+  `${statefulSet.name}-${podID}.${service.name}.${service.namespace}.svc` certificate subjects,
+  the Service CA controller _must_ observe creation of StatefulSet objects.
+  OTOH it does not actually have to observe Services, because the same value is available as
+  `${statefulSet.name}-${podID}.${statefulSet.spec.serviceName}.${statefulSet.namespace}.svc`.
+
+### Generate Multi-Service Certificates
```

Comment:
There was no mention in the RFE of using a single certificate for multiple ports. The ask was for certificates to be issued that could included the SAN of multiple services. This would enable a port securing its traffic with such a certificate to accept traffic from multiple services of potentially different types (i.e. internal clients via clusterip, external clients via nodeip). I think this is an entirely reasonable request.

Reply:
The connection is that if the pod is serving those different services on different ports, it’s trivial to implement the listeners on those ports to serve a different certificate for each port (~= different service), using basically any imaginable TLS server implementation.

Only if two or more services are served on a single port, where the application doesn’t inherently know which service the incoming request is coming for, might the application need extra logic, and a recent TLS server implementation to choose the right certificate (which is easy for _modern_ clients that use SNI; without that, the server might have to guess based on an IP address of the client, or something like that, which is problematic).

The only reasons the user would _need_ a multi-service certificate are 1) exposing a single port as multiple services, or 2) legacy software that can’t be modified (and can’t be hidden behind a service proxy?). The user may _prefer_ a single multi-service certificate so that there are fewer certificates to manage, sure.

The intra-StatefulSet traffic and the traffic for external clients is so conceptually different, that it seems very likely to me that the application implementation is already strongly differentiating between the two, and I’d expect it to already support using different certificates for the two services: If the application is designed to ever be deployed directly over the internet, it already needs to support serving an internet-recognized certificate (e.g. signed by one of the CAs from the browser ecosystem) for external traffic, whereas the intra-StatefulSet traffic is probably firewalled off and using internal host names that can’t be signed by a browser-trusted CA.

---

Anyway, creating combined certificates just based on service/StatefulSet annotations + how selectors of these objects map to pods is impossible in general, as described in the PR.

So we would need to either fall back to brittle heuristics (e.g. only consider pods in service/StatefulSet templates, and ignore any manually-{created,relabeled} pods), or move the secret generation requests out of service/StatefulSet annotations completely, maybe to some CR that contains (secret name to manage, a set of services that should be included in that secret); then the users would be responsible for ensuring that that CR’s set of services matches their actual deployment strategies.

Reply:
To link this back to the original issue:

- etcd uses different ports for client-to-service communication (2379) and among-peers communication (2380), and the certificates / keys can be configured separately: https://etcd.io/docs/current/op-guide/security/ `--cert-file` vs. `--peer-cert-file`
- MinIO supports SNI to choose the appropriate certificate depending on what hostname the client is connecting to: https://github.com/minio/operator/blob/570a0a864eb624aa26b5207bf6ae5e4e0cc4c6f9/examples/tenant-with-custom-ca-certs.yaml#L127

So neither of these applications requires generating a single certificate valid for accepting connections for multiple different services.